statistics/histogram.go

```diff
@@ -1054,3 +1056,67 @@ func matchPrefix(row chunk.Row, colIdx int, ad *types.Datum) bool {
 	}
 	return false
 }
+
+type dataCnt struct {
+	data []byte
+	cnt  uint64
+}
+
+func getIndexPrefixLens(data []byte, numCols int) (prefixLens []int, err error) {
+	prefixLens = make([]int, 0, numCols)
+	var colData []byte
+	prefixLen := 0
+	for len(data) > 0 {
+		colData, data, err = codec.CutOne(data)
+		if err != nil {
+			return nil, err
+		}
+		prefixLen += len(colData)
+		prefixLens = append(prefixLens, prefixLen)
+	}
+	return prefixLens, nil
+}
+
+// ExtractTopN extracts topn from histogram.
+func (hg *Histogram) ExtractTopN(cms *CMSketch, numCols int, numTopN uint32) error {
+	if hg.Len() == 0 || cms == nil || numTopN == 0 {
+		return nil
+	}
+	dataSet := make(map[string]struct{}, hg.Bounds.NumRows())
+	dataCnts := make([]dataCnt, 0, hg.Bounds.NumRows())
+	hg.PreCalculateScalar()
+	// Set a limit on the frequency of boundary values to avoid extract values with low frequency.
+	limit := hg.notNullCount() / float64(hg.Len())
+	// Since our histogram are equal depth, they must occurs on the boundaries of buckets.
+	for i := 0; i < hg.Bounds.NumRows(); i++ {
+		data := hg.Bounds.GetRow(i).GetBytes(0)
+		prefixLens, err := getIndexPrefixLens(data, numCols)
+		if err != nil {
+			return err
+		}
+		for _, prefixLen := range prefixLens {
+			prefixColData := data[:prefixLen]
+			_, ok := dataSet[string(prefixColData)]
+			if ok {
+				continue
+			}
+			dataSet[string(prefixColData)] = struct{}{}
+			res := hg.BetweenRowCount(types.NewBytesDatum(prefixColData), types.NewBytesDatum(kv.Key(prefixColData).PrefixNext()))
```

Comment:
Would it be more accurate to directly use `Repeat` as row count of bucket's upper bound?

Reply:
The `repeats` is only for the whole index, not its prefix, so it is more uniform and the result is also `Repeat` for whole index.